apps/builder/app/builder/features/props-panel/use-props-logic.ts

```diff
@@ -1,156 +1,240 @@
-    type: "string",
-    value: `${propValue?.defaultValue ?? ""}`,
-  };
+  if (meta.type === "number") {
+    return {
+      type: "number",
+      value: meta.defaultValue ?? 0,
+    };
+  }
 
-  if (propValue?.control === "boolean") {
-    typedValue = {
+  if (meta.type === "boolean") {
+    return {
       type: "boolean",
-      value: propValue.defaultValue ?? false,
+      value: meta.defaultValue ?? false,
     };
   }
 
-  if (propValue?.control === "number") {
-    typedValue = {
-      type: "number",
-      value: propValue.defaultValue ?? 0,
+  if (meta.type === "string[]") {
+    return {
+      type: "string[]",
+      value: meta.defaultValue ?? [],
     };
   }
 
-  return typedValue;
+  // If undefined is returned,
+  // we will not add a prop in storage until we get an onChange from control.
+  //
+  // User may have this experience:
+  //   - they added a prop but didn't touch the control
+  //   - they closed props panel
+  //   - when they open props panel again, the prop is not there
+  //
+  // We want to avoid this if possible, but for some types like "asset" we can't
+  return undefined;
+};
+
+const getStartingProp = (
+  instanceId: Instance["id"],
+  meta: PropMeta,
+  name: string
+): Prop | undefined => {
+  const value = getStartingValue(meta);
+  return value && { id: nanoid(), instanceId, name, ...value };
+};
+
+const getDefaultMetaForType = (type: Prop["type"]): PropMeta => {
+  if (type === "string") {
+    return { type: "string", control: "text", required: false };
+  }
+
+  if (type === "number") {
+    return { type: "number", control: "number", required: false };
+  }
+
+  if (type === "boolean") {
+    return { type: "boolean", control: "boolean", required: false };
+  }
+
+  if (type === "asset") {
+    return { type: "string", control: "file-image", required: false };
+  }
+
+  if (type === "string[]") {
+    throw new Error(
+      "A prop with type string[] must have a meta, we can't provide a default one because we need a list of options"
+    );
+  }
+
+  throw new Error(`Usupported data type: ${type}`);
 };
 
-type UsePropsLogic = {
+type UsePropsLogicInput = {
   props: Prop[];
-  selectedInstance: Instance;
-  updateProps: (update: Prop) => void;
+  instance: Instance;
+  updateProp: (update: Prop) => void;
   deleteProp: (id: Prop["id"]) => void;
 };
 
-const getPropFromMetaProps = (
-  instanceId: Instance["id"],
-  metaProps: WsComponentPropsMeta["props"],
-  name: string
-) => {
-  const metaPropValue = metaProps[name];
-  if (metaPropValue === undefined) {
-    return undefined;
-  }
-  return {
-    id: nanoid(),
-    instanceId,
-    name,
-    ...getValueFromPropMeta(metaPropValue),
-  };
+type UsePropsLogicOutput = {
+  /**
+   * Initial (not deletable) props
+   */
+  initialProps: PropAndMeta[];
+  /**
+   * Optional props that were added
+   */
+  optionalProps: PropAndMeta[];
+  /**
+   * Remaining props that can be added
+   */
+  remainingProps: NameAndLabel[];
```

Comment:
remaining sounds like its complete list of the entire remaining properties, if I understand correctly, its the properties that user has manually added? I think there is a better name once I understand the meaning of it

Reply:
It's the complete list of the remaining props still available to add

Reply:
that makes sense then, better to clarify this in the comment, I was not sure